dashboard-api/internal/platform/server/middleware/auth/auth_test.go

```diff
@@ -0,0 +1,144 @@
+package auth
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/json"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/gin-gonic/gin"
+	"github.com/go-playground/assert/v2"
+	"gopkg.in/square/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+)
+
+func TestEnsureValidTokenMiddleware(t *testing.T) {
+	audience := "aud"
+	jwk := GenerateJWK(t)
+	testServer := SetupTestJWKServer(t, jwk)
+	defer testServer.Close()
+
+	t.Run("invalid token", func(t *testing.T) {
+		token := BuildJWTForTesting(
+			t,
+			jwk,
+			testServer.URL,
+			"subject",
+			[]string{"test"},
+		)
+		req, _ := http.NewRequest(http.MethodGet, "", nil)
+		req.Header.Set("Authorization", "Bearer "+token)
+		resp := httptest.NewRecorder()
+		c, _ := gin.CreateTestContext(resp)
+		c.Request = req
+		EnsureValidTokenMiddleware(testServer.URL, audience)(c)
+		assert.Equal(t, resp.Code, 401)
+	})
+	t.Run("valid token", func(t *testing.T) {
+		token := BuildJWTForTesting(
+			t,
+			jwk,
+			testServer.URL,
+			"subject",
+			[]string{audience},
+		)
+		req, _ := http.NewRequest(http.MethodGet, "", nil)
+		req.Header.Set("Authorization", "Bearer "+token)
+		resp := httptest.NewRecorder()
+		c, _ := gin.CreateTestContext(resp)
+		c.Request = req
+		EnsureValidTokenMiddleware(testServer.URL, audience)(c)
+		assert.Equal(t, resp.Code, 200)
+	})
+
+}
+
+// GenerateJWK generates a JWK for testing purposes
+// TODO: Move to internal/testhelpers, currently not possible due to cycle
+// import
+func GenerateJWK(t *testing.T) *jose.JSONWebKey {
+	t.Helper()
+
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatal("failed to generate private key")
+	}
+
+	return &jose.JSONWebKey{
+		Key:       privateKey,
+		KeyID:     "kid",
+		Algorithm: string(jose.RS256),
+		Use:       "sig",
+	}
+}
+
+// SetupTestJWKServer creates a fake server to respond to JWK requests
+// TODO: Move to internal/testhelpers, currently not possible due to cycle
+// import
+func SetupTestJWKServer(
```

Comment:
This seems like a great thing to have for tests, makes it all very easy to test fully!

Reply:
This was taken from Auth0's library so cant take all the credit